### ISSUE TITLE
ObjectBox windows-x64.dll can crash when BoxStore is closed

### ISSUE THREAD

**Issue by MrSlateZB** [objectbox/objectbox-java#723]
ObjectBox version - 2.3.4
Reproducibility - Sometimes. Our tests often need to be ran multiple times for them to be successful. Often they crash with no message / sometimes with the following:

> A fatal error has been detected by the Java Runtime Environment:
> 
> EXCEPTION_ACCESS_VIOLATION (0xc0000005) at pc=0x00007ffbd26b61b2, pid=12764, tid=0x0000000000004260
> JRE version: OpenJDK Runtime Environment (8.0_152-b01) (build 1.8.0_152-release-1343-b01)
> Java VM: OpenJDK 64-Bit Server VM (25.152-b01 mixed mode windows-amd64 compressed oops)
> Problematic frame:
> C  [objectbox-jni-windows-x64.dll+0x461b2]
>
> Failed to write core dump. Minidumps are not enabled by default on client versions of Windows
> 
> An error report file with more information is saved as:
> C:\android\my-app\Common\App\hs_err_pid18348.log
> 
> If you would like to submit a bug report, please visit:
>   http://bugreport.java.com/bugreport/crash.jsp
> The crash happened outside the Java Virtual Machine in native code.
> See problematic frame for where to report the bug.

I've attached the associated file here. The relevant stack trace seems to be:

```
Java frames: (J=compiled Java code, j=interpreted, Vv=VM code)
j  io.objectbox.BoxStore.nativeDelete(J)V+0
j  io.objectbox.BoxStore.close()V+189
j  io.objectbox.BoxStore.finalize()V+81
J 4224 C1 java.lang.ref.Finalizer.runFinalizer(Lsun/misc/JavaLangAccess;)V (62 bytes) @ 0x0000000003f6c1d4 [0x0000000003f6bc40+0x594]
J 4223 C1 java.lang.ref.Finalizer.access$100(Ljava/lang/ref/Finalizer;Lsun/misc/JavaLangAccess;)V (6 bytes) @ 0x0000000003c0d09c [0x0000000003c0d040+0x5c]
j  java.lang.ref.Finalizer$FinalizerThread.run()V+45
v  ~StubRoutines::call_stub
```

[hs_err_pid18348.log](https://github.com/objectbox/objectbox-java/files/3302963/hs_err_pid18348.log)

**Comment by greenrobot-team** [comment]
The log file also contains this:
```
Internal exceptions (10 events):
Event: 6.584 Thread 0x00000000034cd800 Exception <a 'java/lang/NoSuchFieldError': __boxStore> (0x000000066d985850) thrown at [D:\src\AOSP-openjdk-cygwin\jdk8u\hotspot\src\share\vm\prims\jni.cpp, line 2599]
```

`'java/lang/NoSuchFieldError': __boxStore` indicates that the .class files (bytecode) were not properly transformed by the ObjectBox Gradle plugin. Details on this are explained in the docs.
https://docs.objectbox.io/relations#initialization-magic

What kind of tests are these? Android local unit tests? Java unit tests? Maybe you can solve this by explicitly adding the fields the bytecode transformer would add (see linked docs above).

**Comment by MrSlateZB** [comment]
These are Java unit tests running on my PC. We don't use ToOne or ToMany relations in any of our entities. Will that still apply to us then?

**Comment by greenrobot-team** [comment]
Probably not then. Try to compare your test setup against our example.
https://github.com/objectbox/objectbox-examples/blob/master/android-app/src/test/java/io/objectbox/example/AbstractObjectBoxTest.java

**Comment by greenrobot** [comment]
Is there any chance you could post the logs before the crash?

**Comment by MrSlateZB** [comment]
Our base test mirrors the abstract object box test exactly.

We did some additional digging on our end and found that we had some data subscriptions that weren't being closed in our unit tests explicitly. Closing them has stopped this error (and others) from showing up intermittently. 

If this type of error is what you'd expect from that, then we can close this ticket.

**Comment by greenrobot** [comment]
We'd prefer ObjectBox not to crash even if used "wrongly". If you have some additional info (e.g. logs or details/code fragments on the data subscription) that'd be great...

**Comment by MrSlateZB** [comment]
We have a few hundred tests that we were running so unfortunately I can't provide you with any specifics. To give more specifics though we use a modified version of RxQuery where we fixed (https://github.com/objectbox/objectbox-java/issues/687). 

As mentioned above, we had some tests where we were missing dispose calls on our disposables. 

If I had to guess what was the culprit, however, in some of our tests we were using BoxStore#subscribe(Class<T> forClass). We were not applying weak() or any of the other operators to the SubscriptionBuilder. So I'd expect the subscription was trying to be kept alive despite us closing the box after each test? I'm just speculating at this point.
